pkg/network/domainspec/generators.go

```diff
@@ -254,42 +253,51 @@ type PasstLibvirtSpecGenerator struct {
 	vmi          *v1.VirtualMachineInstance
 }
 
+const (
+	passtLogFile      = "/var/run/kubevirt/passt.log" // #nosec G101
```

Comment:
> - Passt logs are now stored at /var/run/kubevirt/passt.log inside virt-launcher.
>   Previously, Passt logs were appended to virt-laucnher log by listening to Passt
>   process stderr.

Why can't we register `/dev/stderr` instead and avoid the goroutine that handles all this?

Reply:
What is the added value of `/dev/stderr`?
The purpose here is to print the passt log to the virt-launcher log in real-time.

Reply:
The pod stderr and stdout eventually reflected in the pod logs.
Passing`/dev/stderr`to libvirt may save working with a file and enable getting rid of the gorutine, because the content that passed to it would reflect in the pod logs.

Reply:
The pod stderr and stdout eventually reflected in the pod logs.
Passing`/dev/stderr`to libvirt may save working with a file and enable getting rid of the gorutine, because the content that passed to it would reflect in the pod logs.

Reply:
Note that `/dev/stderr` shouldn't really be used as log file -- it's a special file, and passt handles size limits and rotation of log files (this was added specifically for KubeVirt, because there's no system logger available), so it needs to truncate the file. The maximum size is 1 MiB by default, by the way.

Reply:
passt has `-e / --stderr`, but libvirt won't pass it (of course -- they are both daemons). You can avoid specifying a log file altogether and, if there's no system logger, messages won't be written anywhere. You can _not_ get standard output or standard error (which are intended for human agents) of passt via libvirt, because, once libvirt starts it, the controlling terminal is detached.

I'm a bit surprised as I implemented `-l / --log-file` specifically for KubeVirt, even though it's certainly not the solution I would have suggested -- system loggers enable, quoting [Wikipedia](https://en.wikipedia.org/wiki/Syslog), "separation of the software that generates messages, the system that stores them, and the software that reports and analyzes them". On this topic, note that passt has no control over the size of libvirt's log files, so it would be rather concerning to let passt use the same file (necessarily, without limitation on the size).

Reply:
@zippy2 IIRC you were working on the log file handling for passt. Can you please take a look at this discussion? Thanks!

Reply:
I checked and kubevirt is not using a file to redirect logs to stderr/stdout of the container [1], it uses stderr of libvirt and passing it the container stdout.

[1] https://github.com/kubevirt/kubevirt/blob/583ab2daeb923f880c316d3ce567b40c398ae32d/pkg/virt-launcher/virtwrap/util/libvirt_helper.go#L229-L246

Reply:
Oh, I didn't know, thanks for pointing that out. So also merging passt's logs with libvirt's logs (which I would strongly advise against) wouldn't help in any case...

Reply:
@sbrivio-rh kubevirt indeed used the `-e` option till now.
It is possible that libvirt will forward passt's stderr to libvirt's stderr?

Reply:
Per se, it's doable, but that would need 1. a very questionable change in libvirt, 2. a new libvirt release, 3. another questionable change in passt (about logging to stderr only).

Expanding on 1., right now the libvirt integration is very simple and robust: `virtqemud` starts passt and doesn't need to keep track of it in any way. If the guest terminates, passt will terminate. If the guest loses connectivity, libvirt will get a QMP event and spawn a new instance of passt, so that QEMU automatically reconnects (that's another feature we added with KubeVirt in mind).

If, however, `virtqemud` terminates, passt won't terminate, which is fundamental because one might want to restart libvirt while a guest is running -- see also https://bugzilla.redhat.com/show_bug.cgi?id=1597326#c2 for the same case around dnsmasq. This is possible because `virtqemud` detaches from passt after starting it. If stderr has to be kept open, this can't be done. So by doing this we would introduce some fundamental issues that are quite complicated to work around later.

On top of this, using stderr from libvirt instead of its (configurable) logging facilities looks like an abuse to me. According to https://libvirt.org/logging.html, stderr can be used only if running _in foreground_. While it seems to work with `cmd.StderrPipe()`, I'm not sure it's intended (because there's no TTY concept at all).

Reply:
The requirement when running in a kubernetes pod is that the logs are accessible through `kubectl logs` [1].
To do so, applications like virt-launcher integrate with other tools and aplications and redirects their logs to stdout, like the example given above with `libvirt`. The same is exercised with the previous passt implementation when it was not fully integrated into libvirt (current state which this PR comes to change).

Now, when using a pipe to redirect stderr/stdout, no file is used and therefore, per my understanding, no EOF is ever reached. The stream will constantly block, waiting for data on the channel.
However, when a file is used, EOF will be reached and therefore we will need to somehow keep track of the current position and re-scan from there. The mechanism to do that is not given, we will have to implement it and manage it. It becomes unclear what happens if the one writing on the log keeps it rotating, as this knowledge is not known to the reader side.
Bottom line, we will need a stream here, not a file, to make this work in a sustainable way without building too much logic over this.

Without getting into the details of "how", we need a way to either allow passt to communicate its logging through libvirt, expose an endpoint that streams through it the logs or just dumps it all on directly to stderr/stdout as we originally hoped to have in this thread. Dumping to stderr directly is not really ideal, as it would have been better to sync the logs and not mess them up (e.g. collision) on output (currently libvirt handles that).

I think that the best we can do now is not to attempt and redirect the logs at all, just to read the logs on virt-launcher termination and provide documentations on how to fetch the passt logs directly from the pod.
Unfortunately this will not integrate well with `kubectl logs` but this seems a good middle ground for now, until we find a better solution.

I guess at some point we will need to detach from this PR and continue the discussion through some other media.

[1] https://kubernetes.io/docs/reference/kubectl/cheatsheet/#interacting-with-running-pods

Reply:
> The requirement when running in a kubernetes pod is that the logs are accessible through `kubectl logs` [1]. To do so, applications like virt-launcher integrate with other tools and aplications and redirects their logs to stdout, like the example given above with `libvirt`. The same is exercised with the previous passt implementation when it was not fully integrated into libvirt (current state which this PR comes to change).

Given that the scope of the logs isn't clearly defined in the Kubernetes documentation, I wonder: does KubeVirt really fetch conceptually equivalent information from kernel logs for other networking modes, about configuration of addresses and routes, and netfilter rules being installed? Are DHCP server messages fetched and included, too? 

That is, I wonder if passt logs really belong to "pod logs".

> Now, when using a pipe to redirect stderr/stdout, no file is used and therefore, per my understanding, no EOF is ever reached.

Strictly speaking, you can have EOFs on all of those (pipes and file descriptors for stderr and stdout) -- but sure, I see the general sense of what you're saying.

> The stream will constantly block, waiting for data on the channel. However, when a file is used, EOF will be reached and therefore we will need to somehow keep track of the current position and re-scan from there. The mechanism to do that is not given, we will have to implement it and manage it. It becomes unclear what happens if the one writing on the log keeps it rotating, as this knowledge is not known to the reader side. Bottom line, we will need a stream here, not a file, to make this work in a sustainable way without building too much logic over this.

Ouch, I feel the pain (even though I still don't understand why we implemented a log file...). But... what you're describing are typical features of a system logger.

If you let applications use typical, appropriate UNIX / Linux facilities (that was the only possibility in passt before we implemented a log file, again specifically for KubeVirt), and ask a system logger to stream or dump, then you don't need to take care of any of this. Neither passt nor libvirt would need to take care of this. They would just send out their messages -- and this separation of roles is a rather relevant security aspect in the design. Standard streams are really meant for interactive usage and terminals instead.